From 41461010eb7c79fee7a9d5f6209accdaac66cc6b Mon Sep 17 00:00:00 2001
From: "H.J. Lu" <hjl.tools@gmail.com>
Date: Sat, 21 Jun 2025 06:52:00 +0800
Subject: [PATCH] elf: Report corrupted group section

Report corrupted group section instead of trying to recover.

	PR binutils/33050
	* elf.c (bfd_elf_set_group_contents): Report corrupted group
	section.

Signed-off-by: H.J. Lu <hjl.tools@gmail.com>
---
 bfd/elf.c | 23 ++++++++++-------------
 1 file changed, 10 insertions(+), 13 deletions(-)

--- binutils.orig/bfd/elf.c	2025-12-16 12:33:34.627390340 +0000
+++ binutils-2.35.2/bfd/elf.c	2025-12-16 12:36:53.274123682 +0000
@@ -3667,8 +3667,18 @@ bfd_elf_set_group_contents (bfd *abfd, a
 	break;
     }
 
+  /* We should always get here with loc == sec->contents + 4.  Return
+     an error for bogus SHT_GROUP sections.  */
   loc -= 4;
-  BFD_ASSERT (loc == sec->contents);
+  if (loc != sec->contents)
+    {
+      /* xgettext:c-format */
+      _bfd_error_handler (_("%pB: corrupted group section: `%pA'"),
+			  abfd, sec);
+      bfd_set_error (bfd_error_bad_value);
+      *failedptr = TRUE;
+      return;
+    }
 
   H_PUT_32 (abfd, sec->flags & SEC_LINK_ONCE ? GRP_COMDAT : 0, loc);
 }
